src/lib.rs

```diff
@@ -1,718 +1,923 @@
+    /// packet, until the MSB is set.
+    GlobalTimestamp2,
 }
 
-/// Tries to parse an ITM packet from the start of the given buffer
-fn parse(input: &[u8]) -> Result<Packet, Either<Error, NeedMoreBytes>> {
-    let header = input.first().cloned().ok_or(Either::Right(NeedMoreBytes))?;
+enum HeaderVariant {
+    Packet(TracePacket),
+    Stub(PacketStub),
+}
 
-    match Header::parse(header).map_err(Either::Left)? {
-        Header::Synchronization => {
-            let mut cursor = 1u8;
+pub struct DecoderOptions {
+    /// Whether to keep reading after a (temporary) EOF condition. If
+    /// set and `next` is called on a decoder iterator, `next` will
+    /// never return if unless the EOF condition is eventually resolved.
+    pub ignore_eof: bool,
+}
 
-            loop {
-                match input.get(usize::from(cursor)) {
-                    Some(&0b0000_0000) => {
-                        // still within the synchronization packet
-                        cursor += 1;
-                        continue;
-                    }
-                    Some(&0b1000_0000) if cursor >= 5 => {
-                        //  "Synchronization packet is at least forty-seven 0 bits followed by single 1
-                        //  bit"
-                        // valid synchronization packet
-                        break Ok(Packet::Synchronization(Synchronization { len: cursor + 1 }));
-                    }
-                    Some(_) => {
-                        // malformed packet
-                        break Err(Either::Left(Error::MalformedPacket {
-                            header,
-                            len: cursor,
-                        }));
-                    }
-                    None => {
-                        // need more bytes
-                        break Err(Either::Right(NeedMoreBytes));
-                    }
-                }
-            }
-        }
+impl Default for DecoderOptions {
+    fn default() -> Self {
+        Self { ignore_eof: true }
+    }
+}
 
-        // Overflow packets have no payload
-        Header::Overflow => Ok(Packet::Overflow),
+#[derive(Debug, thiserror::Error)]
+enum DecoderErrorInt {
+    #[error("Buffer failed to read from source: {0}")]
+    Io(#[from] std::io::Error),
+    #[error("EOF encountered")]
+    Eof,
+    #[error("untars")]
+    MalformedPacket(#[from] MalformedPacket),
+}
 
-        Header::Instrumentation { port, size } => {
-            let mut buffer = [0; 4];
+#[derive(Debug, thiserror::Error)]
+pub enum DecoderError {
+    #[error("I/O error: {0}")]
+    Io(#[from] std::io::Error),
+    #[error("A malformed packet was encountered: {0}")]
+    MalformedPacket(#[from] MalformedPacket),
+}
 
-            let usize = usize::from(size);
-            if input.len() > usize {
-                buffer[..usize].copy_from_slice(&input[1..=usize]);
+struct Buffer<R>
+where
+    R: Read,
+{
+    reader: R,
+    buffer: BitVec,
```

Comment:
We must work bit-by-bit to handle synchronization, in which case the incoming trace stream must be bye-realigned.